---
title:  "<big><u>IMAP-PART 01</u></big><br><br>Preparing for Microbiome Data Analysis with Snakemake Workflows<br><br>"
subtitle: |
  | ![](images/imap_part_01.png)<br>
  | Integrated Microbiome Analysis Pipelines<br>( IMAP-PART 01 ) <br><br><small>[Latest GitHub-Repo](https://github.com/tmbuza/imap-basic-requirements/)</small>
date:
- Updated on `r Sys.Date()`
author: Teresia Mrema-Buza
output: 
  html_document: 
css:
- styles.css
csl: "library/apa.csl"
bibliography:
- library/references.bib
- library/imap.bib
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-basic-requirements"
description: |
  Minimal summary describing features for getting started with microbiome data analysis.
---


<div id="main-container">

<!-- Google Fonts: https://fonts.google.com/?preview.text=IMAP%20template&preview.text_type=custom-->
<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Anton" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@100;300;400;500;700,900&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Oswald:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Merriweather:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@100;200;300;400;700&display=swap" rel="stylesheet">

<span class="fas fa-my-custom-icon"></span>


```{r echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  cache = FALSE,
  comment = NA,
  fig.path='./figures/',
  fig.show='asis',
  dev = 'png',
  fig.align='center',
  out.width = "70%",
  fig.width = 7,
  fig.asp = 0.7,
  fig.show = "asis"
)

library(tidyverse, suppressPackageStartupMessages())
library(schtools)
```

<br>

# Project Overview

## Current project tree
```{bash tree, include=FALSE}
bash workflow/scripts/tree.sh
```

```{bash treeimg}
cat images/project_tree.txt
```

<br><br>

##  Current snakemake workflow
```{bash ruledag, include=FALSE}
bash workflow/scripts/rules_dag.sh
```

<img src="dags/rulegraph.svg">


<br><br>


<!-- ############################# -->
<!-- # Main Subjects Starts here -->
<!-- ############################# -->

# Analysis Software
```{r child='workflow/scripts/software.Rmd'}
```

<br><hr width=100%><br>

# Sample Metadata
<!-- ![](images/metadata.png) -->

```{r child='workflow/scripts/metadata.Rmd'}
```

<br><hr width=100%><br>

# Quality Control Tools

```{r child='workflow/scripts/preprocess_tools.Rmd'}
```

<br><hr width=100%><br>

# Sequencing Data

```{r child='workflow/scripts/sequencing_data.Rmd'}
```

<br><hr width=100%><br>

# Bioinformatics Pipelines

```{r child='workflow/scripts/bioinfo_pipelines.Rmd'}
```

<br><hr width=100%><br>

# References Databases
- Example of reference databases for `Mothur` pipeline.
  - Alignment references
  - Classifiers
  - Taxonomy references
  - Mock references

<br><hr width=100%><br>

# Mapping Files

<!-- ```{r child='workflow/scripts/mapping_files.Rmd'}``` -->

## Mothur sample mapping file
The mapping files are required to direct the pipeline where to look for the files containing the sequencing data.

- The format of the mapping files for `mothur` and `QIIME2` pipelines is slightly different. 
- We will demonstrate how to prepare both.
- Each mapping file will contain only the files that are parsed by bioinformatics analysis. 
- For more detail see Mothur [MiSeq SOP](https://mothur.org/wiki/miseq_sop/).

<br>

## Mothur design file
- Mothur pipeline expects the design (metadata) file to have column headers. 
- We will extract only the desired number of samples.
- The first column header should be **group**.
- For more detail see Mothur [MiSeq SOP](https://mothur.org/wiki/miseq_sop/).


<br><br><hr width=100%><br><br>


# Main Subjects Continue here

<br><br><hr width=100%><br><br>

```{block imaprepos, echo=TRUE}
# Related work

> Here we show the IMAP Repos integrated with Snakemake and GitHub Actions!

<br> 

| Repo | Description | Summary |
|--------------|----------------------------------|--------------|
| [IMAP-PART1](https://github.com/tmbuza/imap-basic-requirements/) | Preparing for microbiome data analysis | [GH-Pages](https://tmbuza.github.io/imap-basic-requirements/) |
| [IMAP-PART2](https://github.com/tmbuza/imap-sample-metadata/) | Exploring and profiling microbiome sample metadata | [GH-Pages](https://tmbuza.github.io/imap-sample-metadata/) |
| [IMAP-PART3](https://github.com/tmbuza/imap-bioinformatics-analysis/) | Bioinformatics analysis of microbiome data | [GH-Pages](https://tmbuza.github.io/imap-bioinformatics-analysis/) |
| [IMAP-PART4](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART5](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART6](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART7](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
| [IMAP-PART8](https://github.com/tmbuza//) |  | [GH-Pages](https://tmbuza.github.io//) |
```

<br>

## Citation
Please consider citing the [iMAP article](https://rdcu.be/b5iVj) (see below) if you find any part of the IMAP practical user guides helpful in your microbiome data analysis.

  Buza, T. M., Tonui, T., Stomeo, F., Tiambo, C., Katani, R., Schilling, M., â€¦ Kapur, V. (2019). iMAP: An integrated bioinformatics and visualization pipeline for microbiome data analysis. BMC Bioinformatics, 20. https://doi.org/10.1186/S12859-019-2965-4

  <br>

> Help to increase awareness of freely available microbiome data analysis tools.  
> Current [iMAP manuscript dimensions](https://badge.dimensions.ai/details/id/pub.1117740326)


<br><br><hr width=100%><br><br>

# References
::: {#refs}
:::

<br><br>

# Appendix

## Static Snakemake report {#smkreport}
> The interactive snakemake html report can be viewed by opening the `report.html` using any compartible browser. You will be able to explore the workflow and the associated statistics. You will also be able to close the left bar to get a better wider view of the display.

```{bash smkrpt, include=FALSE}
snakemake --unlock;
snakemake --report report.html;
hti -H report.html -o images/smkreport;
```

```{bash staticsmk, include=FALSE}
snakemake --report report.html
bash workflow/scripts/smk_html_report.sh
```

<img src="images/smkreport/screenshot.png">

<br><br>

## Troubleshooting
<ol>
  <li>CiteprocXMLError: Missing root element</li>
    <ul>
      <li>Maybe the CSL file is empty. Some examples of citation style language are available on Github [@CSL2023].</li>
      <li></li>
    </ul>
  <!-- <li>Error from srapath: Error: RC(rcNS,rcNoTarg,rcValidating,<br>rcConnection,rcNotFound)</li>
    <ul>
      <li>This happened when downloading SRA reads from online and it seems to be related to connection issues. Not clear what causes this error but keeping retrying sometimes helps.</li>
      <li>Alternatively, you can download the data using the NCBI-SraToolkits (fasterq-dump is the fastest).</li>
    </ul> -->
</ol>


<br><br>

</div> <!-- end main-container -->


<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`. <br><br>
Github_Repo: `r rmarkdown::metadata$github_repo`.  
Repo compiled and maintained by: [`r rmarkdown::metadata$author`](mailto:`r rmarkdown::metadata$email_address`).  


</div><br><br>